Adding Dynamic Load Aggregation Scheme

mitchute opened the pull request:
Pull request overview
---------------------
- Summary:
This pull request introduces a load aggregation scheme to GHEDesigner, enhancing the efficiency of the hourly simulation process while maintaining accuracy. The proposed methodology is based on Claesson and Javed (2012) [1] and Mitchell and Spitler (2019) [2] papers, allowing for faster simulations without compromising reliability.

- PR resolves (GHEDesginer Limitations number 2)
"An hourly simulation is available, but it doesn't make use of load aggregation, so is very slow."

- Key Changes:
Load Aggregation Method: Implemented load aggregation technique outlined in source papers to streamline simulation by aggregating loads within intervals.
Faster Hourly Simulation: The new approach significantly reduces computational overhead, enabling faster results while maintaining accuracy.
Improved Design Workflow: Users can now perform hourly simulations while finding borefield design, eliminating the need for separate simulations, first Hybrid and then Hourly. However, finding design in Hybrid is still the fastest, and gives reliable results in my experience.
Enhanced Performance: GHEDesigner's overall performance is improved, making it suitable for larger simulations and complex heat exchange systems.

- Expected Benefits:
Reduced Computational Time: Load aggregation should significantly decrease the time required for hourly simulations.
Scalability: GHEDesigner will be able to handle larger and more complex systems with improved performance.
Consistent Results: Despite speed improvements, the methodology ensures reliable results for design and analysis with minimum absolute error.

- Comparison between Methods:
This section presents a comprehensive comparison of simulation methods employed in the study. Specifically, the performance of the Hybrid and Load Aggregated Hourly methods is assessed relative to the Hourly simulation, which serves as the standard reference.

A 30-year simulation on Atlanta Office building with 156 boreholes is conducted to evaluate the efficacy of each method in calculating maximum and minimum Hourly Peak Extraction Fluid Temperatures (HPEFTs), alongside the computational efficiency measured in terms of simulation times. 

The error is calculated as the difference between hourly simulation and load aggregated hourly simulation. I can be observed for the 30 years of simulation, that stands to have a maximum absolute value of 0.2 °C.

![Error_30_year_Simulation](https://github.com/BETSRG/GHEDesigner/assets/58135460/641b32b9-dbdd-434b-b441-fbb4fa9b31c4)

The error profile is similar as the 1 year simulation getting repeated each year for this load, as shown below:

![Error_1_year_Simulation](https://github.com/BETSRG/GHEDesigner/assets/58135460/44552bca-61dc-44f5-9470-3b68e052a889)

As the hourly load aggregation depends on expansion rate and number of bins at each level, we followed the suggestions from Mitchell and Spitler paper, where they optimized the ideal expansion rate could be 1.62 or the Golden ratio [2].

For the first 48 hours of the 30 years of simulation, the aggregated simulation and hourly simulation predicts almost the same temperature:

![First 48 hours](https://github.com/BETSRG/GHEDesigner/assets/58135460/39e297ec-8d5c-461a-95b7-e32dd61331b2)

And for the last 48 hours of the simulation also gives pretty accurate results:

![last 48 hours](https://github.com/BETSRG/GHEDesigner/assets/58135460/7e536559-d5c6-475f-8be7-6119764e0e4e)

Upon scrutinizing the maximum and minimum HPEFTs, the Load Aggregation scheme exhibits a maximum absolute error of 0.024°C and 0.004°C, respectively. Notably, these deviations remain within acceptable limits, especially when compared to the Hybrid timestep method. The maximum absolute error for the whole hourly simulation period is 0.2°C.

![Simulation_absolute_error](https://github.com/BETSRG/GHEDesigner/assets/58135460/23caf2b5-dd8e-4921-b197-6b8504599282)

Regarding simulation time, the Load Aggregation method showcases a remarkable improvement, completing the simulation approximately 106 times faster than the conventional Hourly method. However, it is important to note that the Hybrid timestep method outperforms both, achieving a speed enhancement of 15089 times compared to the Hourly method. This efficiency disparity arises from the Hybrid method's monthly simulation frequency as opposed to the hourly frequency of the other methods, resulting in difference in array sizes. The 30-year hourly simulation required approximately 16 minutes to complete, while the Load Aggregation method accomplished the similar task in just 9 seconds.

![Simulation_time_fraction](https://github.com/BETSRG/GHEDesigner/assets/58135460/1f1e2b43-a466-457e-9d4c-cacb4e0ea2a4)

It is worth to note that, the higher is the simulation duration, the load aggregation becomes more time saving. For example, when a 150 hours of Hourly simulation takes about 9 hours and 30 minutes, it takes only 34 seconds with load aggregation scheme, saving 99.9% of computational time. However a 1 year of simulation, takes less than a second for both hourly and load aggregation schemes, where the time saved by aggregation is only 67 %. We have compared the simulation time, and Root Mean Square Error (RMSE) for the whole simulation period for a number of simulation duration, and found the following results:

![Time_Saved](https://github.com/BETSRG/GHEDesigner/assets/58135460/3f61dc48-5355-4ed8-87cd-44cc1a2feb6b)

Thus the simulation time expands exponentially with higher simulation duration in such calculations. The computational complexity of determining the heat pump entering fluid temperature can be expressed in big O notation as O(n!), where 'n' represents the number of time steps utilized in the simulation.

The following chart is presented for a better visualization:

![Simulation_Times](https://github.com/BETSRG/GHEDesigner/assets/58135460/28e6736f-6128-4e5f-bdbb-62ccc4cc7d8f)


For a more detailed comparison with numerical values, please refer to the following table:

![table_comparison](https://github.com/BETSRG/GHEDesigner/assets/58135460/9f7ca7db-cf3a-43c1-b5d7-d14f2bda57ad)




- Request for Review:
Feedback, suggestions, and testing are requested from the development team and community members to ensure the stability and effectiveness of the load aggregation feature.

Thank you for your contributions!
Ishraque Zaman Borshon

References:
[1] Claesson, Johan, and Saqib Javed. "A load-aggregation method to calculate extraction temperatures of borehole heat exchangers." Ashrae Transactions 118, no. 1 (2012): 530-540.
[2] Mitchell, Matt S., and Jeffrey D. Spitler. "Characterization, testing, and optimization of load aggregation methods for ground heat exchanger response-factor models." Science and Technology for the Built Environment 25, no. 8 (2019): 1036-1051.

### Checklist
- [x] Documentation added/updated
- [x] CI status: all green or justified

ishraque2008 committed:
Adding Dynamic Load Aggregation scheme.

mitchute commented:
Thanks, @ishraque2008. Please add a basic description of the work to the "Pull request overview" above. We also will need a unit test to exercise the new code. I'll also likely provide some comments related to portions of the code soon.

mitchute commented on ghedesigner/enums.py:
```diff
@@ -16,6 +16,7 @@ class DoubleUTubeConnType(Enum):
 class TimestepType(Enum):
     HOURLY = auto()
     HYBRID = auto()
+    DYNAMIC = auto()
```
You're proposing a new load aggregation scheme, not a timestep method, right? Unless I'm missing something, this should not be included here. Or maybe we also need some other cleanups in general to clarify the usage. Ideally, we should be able to have hourly timesteps and still use any of the load aggregation schemes that have been proposed.

ishraque2008 replied:
Thanks! Removed it, and included load aggregation to the Hourly timestep function.

mitchute commented on ghedesigner/ground_heat_exchangers.py:
```diff
@@ -177,6 +177,77 @@ def _simulate_detailed(
+        # Source: Claesson, J., & Javed, S. (2012). A load-aggregation method to calculate extraction temperatures of
+        # borehole heat exchangers. ASHRAE Transactions, 118(1), 530-540.
+
+    def Simulate_Dynamic_LA(self, g, q_dot_hourly: np.ndarray):
```
Function names should be lower-cased.

ishraque2008 replied:
Done.

mitchute commented on ghedesigner/ground_heat_exchangers.py:
```diff
@@ -177,6 +177,77 @@ def _simulate_detailed(
+        delta_tb = np.zeros(len(q_dot))
+        g_s = g(np.log((v_array[1:] * 3600.0) / ts))  # g-values
+
+        Q_n = np.zeros(q_max * p_q)  # Array holding current load
```
We're not totally consistent with Python naming conventions in GHEDesginer, but I would like to move that way. Following PEP 8, variable names should also be lower-cased. https://peps.python.org/pep-0008

ishraque2008 replied:
Modified the variables to lower case.

mitchute commented on ghedesigner/ground_heat_exchangers.py:
```diff
@@ -337,6 +408,22 @@ def simulate(self, method: TimestepType):
 
             hp_eft, d_tb = self._simulate_detailed(q_dot, t, g)
+
+        elif method == TimestepType.DYNAMIC:
```
As noted previously, this isn't a timestep method, so we may need to do some more work here to make sure things are in the right place.

ishraque2008 replied:
Done.

ishraque2008 committed:
Modifying hourly timestep scheme and variable names.

ishraque2008 committed:
Modifying hourly timestep scheme and variable names.

ishraque2008 committed:
Modifying hourly timestep scheme and variable names.

ishraque2008 committed:
Modifying hourly timestep scheme and variable names.

ishraque2008 committed:
Merge remote-tracking branch 'origin/Load_Aggregation' into Load_Aggregation

ishraque2008 committed:
Merge branch 'main' into Load_Aggregation

ishraque2008 committed:
Merge branch 'main' into Load_Aggregation

ishraque2008 committed:
Merge remote-tracking branch 'origin/Load_Aggregation' into Load_Aggregation

Commit:
Merge branch 'main' into Load_Aggregation

Commit:
Addition of Test and Demo scripts
Added Test and Demo scripts, updating parameters in Demo file to align with latest version.

Commit:
Added a newline at the end of script to avoid flake8 failure.

Commit:
Added Hybrid as default timestep to set_design() function.

mitchute commented:
@ishraque2008 thanks for picking this back up again!

If you'd rather not wait for GitHub to run the tests, which takes 10+ minutes, you can run them locally. After pip installing the requirements (`pip install -r requirements.txt`), you can run the following command to run the tests locally.

```
$ pytest
```

You can also add the `-v` flag to display verbose output, and the `-x` flag to stop after a test has failed.

E.g. `$ pytest -v`, `$ pytest -x` and combined `$ pytest -vx`

Depending on your setup there may be a few other tricks that are needed, so if you run into any issues with that let me know and I can help.

Commit:
Implement hourly simulation method for finding borefield designs and resolve associated test failures.

ishraque2008 commented:
> @ishraque2008 thanks for picking this back up again!
> 
> If you'd rather not wait for GitHub to run the tests, which takes 10+ minutes, you can run them locally. After pip installing the requirements (`pip install -r requirements.txt`), you can run the following command to run the tests locally.
> 
> ```
> $ pytest
> ```
> 
> You can also add the `-v` flag to display verbose output, and the `-x` flag to stop after a test has failed.
> 
> E.g. `$ pytest -v`, `$ pytest -x` and combined `$ pytest -vx`
> 
> Depending on your setup there may be a few other tricks that are needed, so if you run into any issues with that let me know and I can help.


@mitchute Sorry about the long delay to pick up. I'll make sure to be more prompt from now on.
And thanks for the guidance! I've followed your instructions to run the tests. It has been a real-time saver! It would take forever otherwise. Please let me know if there's anything else that needs attention.

mitchute commented on ghedesigner/tests/test_load_aggregation.py:
```diff
@@ -0,0 +1,267 @@
+from ghedesigner.utilities import eskilson_log_times
+
+
+class TestLoadAggregation(GHEBaseTest):
```
It would be much preferred if you could update this test to follow this example.

https://github.com/BETSRG/GHEDesigner/blob/e055caa449285499a58a043961ee27946dcd3ca0/ghedesigner/tests/test_find_design_rectangle.py#L6-L26

The previous tests were quite verbose. The new approach is a much more compact way to do the same thing.

ishraque2008 replied:
Thanks! I fused the two concepts and made the test even more concise.

mitchute commented on ghedesigner/manager.py:
```diff
@@ -804,6 +812,8 @@ def run_manager_from_cli_worker(input_file_path: Path, output_directory: Path) -
         continue_if_design_unmet=continue_if_design_unmet
     )
 
+    timestep = sim_props.get("timestep", "HYBRID").upper()
```
Nice touch. I was not aware of this `dict` function.

mitchute reviewed (commented):
This generally looks really good, but I probably will still do a little more testing.

One thing that needs to be changed is to update the following description in the simulation schema.

https://github.com/BETSRG/GHEDesigner/blob/441419b93faa28379d54afde6388c35032de5f67/ghedesigner/schemas/simulation.schema.json#L36

mitchute commented:
From the PR description, [above](https://github.com/BETSRG/GHEDesigner/pull/81#issue-1812048917)

> Reduced Computational Time: The load aggregation scheme is expected to significantly reduce the computational time required for hourly simulations, making GHEDesigner more efficient for various projects.

> Scalability: With the load aggregation method, GHEDesigner should be able to handle larger and more complex heat exchange systems with improved performance.

> Consistent Results: Despite the speed improvement, the methodology still maintains a high level of accuracy, ensuring reliable results for design and analysis purposes.

It would be fantastic if you could give a few statistics, plots of differences in simulation times and design outputs, or something else to help see what improvements this may or may not offer.

ishraque2008 commented:
> This generally looks really good, but I probably will still do a little more testing.
> 
> One thing that needs to be changed is to update the following description in the simulation schema.
> 
> https://github.com/BETSRG/GHEDesigner/blob/441419b93faa28379d54afde6388c35032de5f67/ghedesigner/schemas/simulation.schema.json#L36

@mitchute Thanks for your inputs. Let me address all the suggestions.

Commit:
Updated description of timestep options in schema to include 'HOURLY' alongside 'HYBRID', allowing for more flexible simulation configurations

Commit:
Refactoring test file for improved efficiency and readability.

mitchute committed:
make test output directory names unique

ishraque2008 commented:
> From the PR description, [above](https://github.com/BETSRG/GHEDesigner/pull/81#issue-1812048917)
> 
> > Reduced Computational Time: The load aggregation scheme is expected to significantly reduce the computational time required for hourly simulations, making GHEDesigner more efficient for various projects.
> 
> > Scalability: With the load aggregation method, GHEDesigner should be able to handle larger and more complex heat exchange systems with improved performance.
> 
> > Consistent Results: Despite the speed improvement, the methodology still maintains a high level of accuracy, ensuring reliable results for design and analysis purposes.
> 
> It would be fantastic if you could give a few statistics, plots of differences in simulation times and design outputs, or something else to help see what improvements this may or may not offer.


> From the PR description, [above](https://github.com/BETSRG/GHEDesigner/pull/81#issue-1812048917)
> 
> > Reduced Computational Time: The load aggregation scheme is expected to significantly reduce the computational time required for hourly simulations, making GHEDesigner more efficient for various projects.
> 
> > Scalability: With the load aggregation method, GHEDesigner should be able to handle larger and more complex heat exchange systems with improved performance.
> 
> > Consistent Results: Despite the speed improvement, the methodology still maintains a high level of accuracy, ensuring reliable results for design and analysis purposes.
> 
> It would be fantastic if you could give a few statistics, plots of differences in simulation times and design outputs, or something else to help see what improvements this may or may not offer.


@mitchute Done. Please let me know if more statistics are required.

ishraque2008 commented:
> > From the PR description, [above](https://github.com/BETSRG/GHEDesigner/pull/81#issue-1812048917)
> > > Reduced Computational Time: The load aggregation scheme is expected to significantly reduce the computational time required for hourly simulations, making GHEDesigner more efficient for various projects.
> > 
> > 
> > > Scalability: With the load aggregation method, GHEDesigner should be able to handle larger and more complex heat exchange systems with improved performance.
> > 
> > 
> > > Consistent Results: Despite the speed improvement, the methodology still maintains a high level of accuracy, ensuring reliable results for design and analysis purposes.
> > 
> > 
> > It would be fantastic if you could give a few statistics, plots of differences in simulation times and design outputs, or something else to help see what improvements this may or may not offer.
> 
> > From the PR description, [above](https://github.com/BETSRG/GHEDesigner/pull/81#issue-1812048917)
> > > Reduced Computational Time: The load aggregation scheme is expected to significantly reduce the computational time required for hourly simulations, making GHEDesigner more efficient for various projects.
> > 
> > 
> > > Scalability: With the load aggregation method, GHEDesigner should be able to handle larger and more complex heat exchange systems with improved performance.
> > 
> > 
> > > Consistent Results: Despite the speed improvement, the methodology still maintains a high level of accuracy, ensuring reliable results for design and analysis purposes.
> > 
> > 
> > It would be fantastic if you could give a few statistics, plots of differences in simulation times and design outputs, or something else to help see what improvements this may or may not offer.
> 
> @mitchute Done. Please let me know if more statistics are required.

Hello @mitchute I just added some more data to the description.

mitchute committed:
Merge branch 'main' into Load_Aggregation

mitchute committed:
update num tests

mitchute committed:
formating

mitchute committed:
simplify simulate function with load agg

mitchute committed:
cleanup, simplification

mitchute committed:
additonal cleanups, more work on tests

mitchute committed:
remove unused arg

mitchute committed:
pruning unnecessary args, further cleanups

mitchute committed:
Fix IF statement for hourly w/wo load agg, fix bh wall temp output var

mitchute committed:
reduce runtime on non load aggregation case

mitchute commented:
Alright, I think I have done as much as I can at this point. I'm going to need to kick it back to you (@ishraque2008) for some further investigation. 

To get here, I have cleaned up the code so we are not passing in duplicated, unnecessary arguments. I have also added another member to the enum for tracking which timestep method is being used, then updated the logic to switch on those enum values instead of the strings that were being passed around. I'm fairly confident all the right code is being hit now once the proper flag is actuated. To my knowledge, I didn't touch any of the new load aggregation code.

However, I think there are still some issues here that warrant further investigation. I looked at the following schemes and sizing periods: Hybrid (20 yr), Hourly with LA (20 yr), Hourly w/o LA (1-8, 10, and 20 yr). See the plots below. (the cases are labeled by the number of months in the sizing period)

In the first plot, we see the GHE ExFT for the Hybrid, Hourly with and w/o LA for 20 years. Over the sizing period, the hourly with LA consistently underpredicts the fluid temp when compared to the straight hourly simulation. In my opinion, this leads to the results of the second plot.

![Screenshot 2024-05-01 at 10 54 06 AM](https://github.com/BETSRG/GHEDesigner/assets/2985979/d3f20cf4-aae9-458b-a2d5-20ac8d440630)

In this plot, we see the total drilling and simulation runtime for the cases noted above. I was expecting the 20-year cases (Hybrid, Hourly w LA, and Hourly w/o LA) to be consistent. However, the case with load aggregation yields a design that is roughly 11% smaller than the other two cases, which in my opinion warrants further investigation.

![Screenshot 2024-05-01 at 9 57 19 AM](https://github.com/BETSRG/GHEDesigner/assets/2985979/ca918f73-a70d-4089-bfac-19c9faf9e41d)

I also found it strange that the design jumps by roughly 2X when changing the sizing period from 10-20 years, but perhaps this is expected given the loads we're using. @jeffreyspitler, feel free to comment on that point if you think something looks off.

My notebook is attached, so let me know if I missed something. Otherwise, I think you might double-check that your LA algorithm is working as expected.

[Compare Results.ipynb.zip](https://github.com/BETSRG/GHEDesigner/files/15179492/Compare.Results.ipynb.zip)

ishraque2008 commented:
> Alright, I think I have done as much as I can at this point. I'm going to need to kick it back to you (@ishraque2008) for some further investigation.
> 
> To get here, I have cleaned up the code so we are not passing in duplicated, unnecessary arguments. I have also added another member to the enum for tracking which timestep method is being used, then updated the logic to switch on those enum values instead of the strings that were being passed around. I'm fairly confident all the right code is being hit now once the proper flag is actuated. To my knowledge, I didn't touch any of the new load aggregation code.
> 
> However, I think there are still some issues here that warrant further investigation. I looked at the following schemes and sizing periods: Hybrid (20 yr), Hourly with LA (20 yr), Hourly w/o LA (1-8, 10, and 20 yr). See the plots below. (the cases are labeled by the number of months in the sizing period)
> 
> In the first plot, we see the GHE ExFT for the Hybrid, Hourly with and w/o LA for 20 years. Over the sizing period, the hourly with LA consistently underpredicts the fluid temp when compared to the straight hourly simulation. In my opinion, this leads to the results of the second plot.
> 
> ![Screenshot 2024-05-01 at 10 54 06 AM](https://private-user-images.githubusercontent.com/2985979/327183601-d3f20cf4-aae9-458b-a2d5-20ac8d440630.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ2MzUzMDAsIm5iZiI6MTcxNDYzNTAwMCwicGF0aCI6Ii8yOTg1OTc5LzMyNzE4MzYwMS1kM2YyMGNmNC1hYWU5LTQ1OGItYTJkNS0yMGFjOGQ0NDA2MzAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MDJUMDczMDAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZGExNzBhMDI2YzJhMjFkNmY1MzVkZjA0NGY2ODEzY2RiM2Y3MDY0MGU2MTMzYzkxNWQ1NzU5YWYzMDdkMGQ0YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.7eetgd6gZvc1MTcxYLIUFiwbgAqhrybW7x22Gwke60Q)
> 
> In this plot, we see the total drilling and simulation runtime for the cases noted above. I was expecting the 20-year cases (Hybrid, Hourly w LA, and Hourly w/o LA) to be consistent. However, the case with load aggregation yields a design that is roughly 11% smaller than the other two cases, which in my opinion warrants further investigation.
> 
> ![Screenshot 2024-05-01 at 9 57 19 AM](https://private-user-images.githubusercontent.com/2985979/327183609-ca918f73-a70d-4089-bfac-19c9faf9e41d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ2MzUzMDAsIm5iZiI6MTcxNDYzNTAwMCwicGF0aCI6Ii8yOTg1OTc5LzMyNzE4MzYwOS1jYTkxOGY3My1hNzBkLTQwODktYmZhYy0xOWM5ZmFmOWU0MWQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MDJUMDczMDAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Njk3ZGUwOWIwNWE3NGI1ZWIwNTYxOGNlNDRjOThjMDBhZGJjM2MyMjc1MmIxNzAyYTZmZjBkNzJkMmE0MmMxNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.QBEexAn0UKyhz7PAzEXDEkXtjylkB7cZuJaRLc51ZiA)
> 
> I also found it strange that the design jumps by roughly 2X when changing the sizing period from 10-20 years, but perhaps this is expected given the loads we're using. @jeffreyspitler, feel free to comment on that point if you think something looks off.
> 
> My notebook is attached, so let me know if I missed something. Otherwise, I think you might double-check that your LA algorithm is working as expected.
> 
> [Compare Results.ipynb.zip](https://github.com/BETSRG/GHEDesigner/files/15179492/Compare.Results.ipynb.zip)


Interesting! I thought everything was working pre

> Alright, I think I have done as much as I can at this point. I'm going to need to kick it back to you (@ishraque2008) for some further investigation.
> 
> To get here, I have cleaned up the code so we are not passing in duplicated, unnecessary arguments. I have also added another member to the enum for tracking which timestep method is being used, then updated the logic to switch on those enum values instead of the strings that were being passed around. I'm fairly confident all the right code is being hit now once the proper flag is actuated. To my knowledge, I didn't touch any of the new load aggregation code.
> 
> However, I think there are still some issues here that warrant further investigation. I looked at the following schemes and sizing periods: Hybrid (20 yr), Hourly with LA (20 yr), Hourly w/o LA (1-8, 10, and 20 yr). See the plots below. (the cases are labeled by the number of months in the sizing period)
> 
> In the first plot, we see the GHE ExFT for the Hybrid, Hourly with and w/o LA for 20 years. Over the sizing period, the hourly with LA consistently underpredicts the fluid temp when compared to the straight hourly simulation. In my opinion, this leads to the results of the second plot.
> 
> ![Screenshot 2024-05-01 at 10 54 06 AM](https://private-user-images.githubusercontent.com/2985979/327183601-d3f20cf4-aae9-458b-a2d5-20ac8d440630.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ2MzUzMDAsIm5iZiI6MTcxNDYzNTAwMCwicGF0aCI6Ii8yOTg1OTc5LzMyNzE4MzYwMS1kM2YyMGNmNC1hYWU5LTQ1OGItYTJkNS0yMGFjOGQ0NDA2MzAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MDJUMDczMDAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZGExNzBhMDI2YzJhMjFkNmY1MzVkZjA0NGY2ODEzY2RiM2Y3MDY0MGU2MTMzYzkxNWQ1NzU5YWYzMDdkMGQ0YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.7eetgd6gZvc1MTcxYLIUFiwbgAqhrybW7x22Gwke60Q)
> 
> In this plot, we see the total drilling and simulation runtime for the cases noted above. I was expecting the 20-year cases (Hybrid, Hourly w LA, and Hourly w/o LA) to be consistent. However, the case with load aggregation yields a design that is roughly 11% smaller than the other two cases, which in my opinion warrants further investigation.
> 
> ![Screenshot 2024-05-01 at 9 57 19 AM](https://private-user-images.githubusercontent.com/2985979/327183609-ca918f73-a70d-4089-bfac-19c9faf9e41d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ2MzUzMDAsIm5iZiI6MTcxNDYzNTAwMCwicGF0aCI6Ii8yOTg1OTc5LzMyNzE4MzYwOS1jYTkxOGY3My1hNzBkLTQwODktYmZhYy0xOWM5ZmFmOWU0MWQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MDJUMDczMDAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Njk3ZGUwOWIwNWE3NGI1ZWIwNTYxOGNlNDRjOThjMDBhZGJjM2MyMjc1MmIxNzAyYTZmZjBkNzJkMmE0MmMxNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.QBEexAn0UKyhz7PAzEXDEkXtjylkB7cZuJaRLc51ZiA)
> 
> I also found it strange that the design jumps by roughly 2X when changing the sizing period from 10-20 years, but perhaps this is expected given the loads we're using. @jeffreyspitler, feel free to comment on that point if you think something looks off.
> 
> My notebook is attached, so let me know if I missed something. Otherwise, I think you might double-check that your LA algorithm is working as expected.
> 
> [Compare Results.ipynb.zip](https://github.com/BETSRG/GHEDesigner/files/15179492/Compare.Results.ipynb.zip)

mitchute commented:
Closing this for now. Feel free to reopen it if you decide to pick it back up.

ishraque2008 committed:
Updating to merge main branch into Load_Aggregation.

ishraque2008 committed:
Fixed Load_Aggregation method for this branch.

ishraque2008 committed:
Cleaning up code

ishraque2008 committed:
Cleaning up code-2

ishraque2008 commented:
Hi @mitchute , thanks again for your earlier review and for pointing out the discrepancy in the hourly load aggregation results. I picked this back up and reworked the implementation, updated the tests/demo path, and adding new figures and comparison data.

As it appears, I have reproduced the "Balanced" and "Imbalanced" loads as described by your thesis, and it seems the results are very close with the Non-aggregated method for a 20 years of simulation. 

For balanced load the comparison looks like having error below 0.1°C.
<img width="1189" height="789" alt="image" src="https://github.com/user-attachments/assets/7537da1a-617e-4d54-94ff-afe77897da94" />

And for imbalanced load it is also good, less than 0.1° C error.
<img width="1189" height="790" alt="image" src="https://github.com/user-attachments/assets/0c8f1294-7d46-42d0-bebd-00c3099124f2" />

However, whenever we take extremely imbalanced load or the "real" building loads, this is not that perfect. For example, for Atlanta office building load (annual cooling-to-heating load ratio is around 23.), it shows error just below 1°C. However, the real picture comparison can be tested with the hybrid load, that shows error above 3°C (Green lines). So the load aggregation is performing better than it acually was.


<img width="1181" height="990" alt="image" src="https://github.com/user-attachments/assets/6b6c4174-af19-4cec-909b-28318c6c96fd" />

But this is not true for all "real" loads, for example for Chicago Office building, the results are a little better than the synthetic loads:

<img width="1189" height="790" alt="image" src="https://github.com/user-attachments/assets/46de0c92-d421-4220-957b-3fad2072dc82" />

While looking at the sizing results, there are some interesting observations with Atlanta office building, which was missing during my previous implementation. I will show this after showing some results first.

When sizing for a balanced load, the load aggregation method yields borehole heights more consistent with the no-load aggregation method than the hybrid approach does.

Balanced load:
<img width="989" height="790" alt="image" src="https://github.com/user-attachments/assets/04b61265-c5cc-4b83-9a9f-e2e10b5c985d" />

And it is followed by the imbalanced load too:
<img width="989" height="790" alt="image" src="https://github.com/user-attachments/assets/43497623-8174-4172-82dc-aa7169fcad59" />


The Chicago office building also shows better alignment than the hybrid load:
<img width="989" height="790" alt="image" src="https://github.com/user-attachments/assets/7dbf0956-bf6c-4862-bd4f-91a8a037df86" />


However, for Atlanta office building load, the load aggregation method will show very confusing result:

<img width="989" height="790" alt="image" src="https://github.com/user-attachments/assets/d7cadeeb-6891-4425-b49a-aa94385ae9ab" />

As you can imagine, this is not because the load aggregation method is not performing as expected, rather because of the other methods that depend on the overall design procedure, including iteration during sizing and also while searching optimized borefield grid. During the field search iterations, both the Hybrid and Hourly without load aggregation methods reached the 18×8 configuration (B = 5.00 m) and, interestingly, both reported an excess temperature of 0.01 °C, making this the apparent near-optimal design point.

However, for the same 18×8_B5.00 configuration, the Hourly Load Aggregation method predicted an excess temperature of −0.31°C instead of 0.01°C. Because the search algorithm uses this excess temperature to guide the next iteration, this small deviation caused the subsequent search path to move in a different direction. As a result, the load aggregation workflow proceeded toward 13×6_B7.08 and 15×7_B6.07, ultimately leading to a less optimal design outcome compared with the Hybrid result.

That said, as shown in the earlier comparison plots, the load aggregation method still tracks the Hourly no-load-aggregation solution much more closely overall than the Hybrid method in terms of transient temperature prediction. Therefore, the discrepancy here appears to be primarily a search-path sensitivity issue near the convergence threshold, rather than a fundamental issue with the load aggregation temperature prediction itself. My working files are messy, thats why I am not sharing them here, but would be happy to share any files you think would help you evaluate better. I tested the method with 7-8 other building loads too (which were used for the validation of GHEDesigner), and found, while simulation shows close alignment, the sizing may not always give the best result with load aggregation method.

Finally, the updated results look much better now: the hourly load aggregation case follows the hourly no-load-aggregation case much more closely, and I am no longer seeing the earlier unexpectedly smaller design result. Sorry again for the long gap on this. I have been in the final stretch of wrapping up my PhD, which pulled me away from revisiting this branch sooner, but I wanted to come back with something substantially improved rather than a partial update. I really appreciate your earlier feedback, and especially your notebook was instrumental in helping me track down where the discrepancy was coming from. 

While working through this, I also started exploring a few follow-up directions around even faster simulation and surrogate prediction of thermal response behavior, and the preliminary results are quite encouraging. I did not want to overload this PR thread, but I would be happy to share them at your convenience, if you are interested.
